service/awsconfig/v5/resource/cloudformation/update.go

```diff
@@ -64,24 +84,109 @@ func (r *Resource) newUpdateChange(ctx context.Context, obj, currentState, desir
 		return cloudformation.CreateStackInput{}, microerror.Mask(err)
 	}
 
-	r.logger.LogCtx(ctx, "debug", "finding out if the main stack should be updated")
+	// We enable/disable updates in order to enable them our test installations
+	// but disable them in production installations. That is useful until we have
+	// full confidence in updating guest clusters. Note that updates also manage
+	// scaling at the same time to be more efficient.
+	if updateallowedcontext.IsUpdateAllowed(ctx) {
+		r.logger.LogCtx(ctx, "debug", "finding out if the main stack has to be updated")
 
-	updateState := cloudformation.UpdateStackInput{}
+		if shouldUpdate(currentStackState, desiredStackState) {
+			r.logger.LogCtx(ctx, "debug", "main stack has to be updated")
 
-	if currentStackState.Name != "" && !reflect.DeepEqual(desiredStackState, currentStackState) {
-		r.logger.LogCtx(ctx, "debug", "main stack should be updated")
+			updateState, err := r.computeUpdateState(customObject, desiredStackState)
+			if err != nil {
+				return cloudformation.CreateStackInput{}, microerror.Mask(err)
+			}
 
-		mainTemplate, err := r.getMainGuestTemplateBody(customObject)
-		if err != nil {
-			return nil, microerror.Mask(err)
+			return updateState, nil
+		} else {
+			r.logger.LogCtx(ctx, "debug", "main stack has not to be updated")
 		}
-		updateState.StackName = aws.String(desiredStackState.Name)
-		updateState.TemplateBody = aws.String(mainTemplate)
-		// CAPABILITY_NAMED_IAM is required for updating IAM roles (worker policy).
-		updateState.Capabilities = []*string{
-			aws.String(namedIAMCapability),
+	} else {
+		r.logger.LogCtx(ctx, "debug", "not computing update state because main stack are not allowed to be updated")
+	}
+
+	// We manage scaling separately because the impact and implications of scaling
+	// is different compared to updates. We can just process scaling any time. We
+	// cannot just process updates at any time and thus have to separate the
+	// management of both primitives. Note that updates also manage scaling at the
+	// same time for more efficiency.
+	if shouldScale(currentStackState, desiredStackState) {
+		r.logger.LogCtx(ctx, "debug", "main stack has to be scaled")
+
+		updateState, err := r.computeUpdateState(customObject, desiredStackState)
+		if err != nil {
+			return cloudformation.CreateStackInput{}, microerror.Mask(err)
 		}
+
+		return updateState, nil
+	} else {
+		r.logger.LogCtx(ctx, "debug", "main stack has not to be scaled")
 	}
 
-	return updateState, nil
+	return cloudformation.UpdateStackInput{}, nil
+}
+
+// shouldScale determines whether the reconciled guest cluster should be scaled.
+// A guest cluster is only allowed to scale in case nothing but the worker count
+// changes. In case anything else changes as well, scaling is not allowed, since
+// any other changes should be covered by general updates, which is a separate
+// step.
+func shouldScale(currentState, desiredState StackState) bool {
```

Comment:
FYI I removed the check for the name change. IMO the name should never change and should also not affect reconciliation. As of now we use the cluster ID and some magic string format which should just be fine to be ignored here.